### PR TITLE
Feature/read install config

### PR DESCRIPTION
**Pull request by henrikstranneheim**
### This PR adds | fixes:

- Added sub routines to read install config and get container part
- Added cli option to specify an install config in analysis
- Added install_config_file parameter in pipeline parameters definition
- Added default option using template
- Added sub routine to get container part from install config
- Added wrapper for getting container_cmds via container_api
- Set container cmds in constants
- Get container cmds from constants
- Added bind paths for recipes and MIP globally
- Added wrapper to all program API:s that use containers
- Changed command_module template to use container wrapper
- Added tests

### How to test:

- Automatic and continuous test suite

### Expected outcome:
- [ ] Installation, unit and integration test suite pass

### Review:
- [ ] Code review
- [ ] Tests pass

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes. If applicable record manual test results in PR header
- [ ] **MINOR** - when you add functionality in a backwards compatible manner. If applicable record manual test results in PR header
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
@jemten This i a start of what we talked about with containers today. We could continue from this one, or merge this one and continue in a new PR whoever has the time (most likely you....)

File: definitions/analyse_parameters.yaml [hunk 1]
```diff
@@ -47,6 +47,13 @@ case_id:
     - mip
   data_type: SCALAR
   type: mip
+install_config_file:
```
Comment (jemten) [inline definitions/analyse_parameters.yaml:50]: Maybe we should name this option differently. I first read it as _install this config file_, like this is the config file we want to install. I realised later that this had nothing with to do with installation. We can discuss it later perhaps

**Review by jemten — APPROVED** [review]
Very nice! A first step towards full container integration

**Comment by henrikstranneheim** [comment]
@jemten and @ramprasadn Test run is on the way on hasta. Please have a look at the code

**Comment by henrikstranneheim** [comment]
Result from test run:
```
[0|0|2615] [henrik.stranneheim@hasta:~] [base] 4s $ cat develop/livingox/analysis/mip_log/2020-11-07/mip_2020-11-07T18\:54\:00.log.status
#jobid	jobname	account	partition	alloccpus	TotalCPU	elapsed	start	end	state	exitcode
721801	fastqc_ar_ACC5339A2	developme+	core	8	03:40:07	00:28:21	2020-11-07T18:54:17	2020-11-07T19:22:38	COMPLETED	0:0
721802	fastqc_ar_ACC5339A3	developme+	core	8	03:45:36	00:29:27	2020-11-07T18:54:17	2020-11-07T19:23:44	COMPLETED	0:0
721803	fastqc_ar_ACC5339A4	developme+	core	8	03:33:01	00:27:28	2020-11-07T18:54:17	2020-11-07T19:21:45	COMPLETED	0:0
721804	bwa_mem_ACC5339A2	developme+	core	36	1-11:00:07	01:18:30	2020-11-07T18:54:17	2020-11-07T20:12:47	COMPLETED	0:0
721805	bwa_mem_ACC5339A2	developme+	core	36	1-10:35:19	01:19:37	2020-11-07T18:54:17	2020-11-07T20:13:54	COMPLETED	0:0
721806	bwa_mem_ACC5339A2	developme+	core	36	1-10:18:30	01:17:10	2020-11-07T18:54:17	2020-11-07T20:11:27	COMPLETED	0:0
721807	bwa_mem_ACC5339A2	developme+	core	36	1-10:35:42	01:17:39	2020-11-07T18:54:17	2020-11-07T20:11:56	COMPLETED	0:0
721808	bwa_mem_ACC5339A3	developme+	core	36	1-14:02:38	01:23:55	2020-11-07T18:54:17	2020-11-07T20:18:12	COMPLETED	0:0
721809	bwa_mem_ACC5339A3	developme+	core	36	1-11:04:37	01:20:44	2020-11-07T18:54:17	2020-11-07T20:15:01	COMPLETED	0:0
721810	bwa_mem_ACC5339A3	developme+	core	36	1-11:17:42	01:21:34	2020-11-07T18:54:17	2020-11-07T20:15:51	COMPLETED	0:0
721811	bwa_mem_ACC5339A3	developme+	core	36	1-11:26:14	01:19:02	2020-11-07T18:54:17	2020-11-07T20:13:19	COMPLETED	0:0
721812	bwa_mem_ACC5339A4	developme+	core	36	1-08:47:22	01:15:47	2020-11-07T18:54:17	2020-11-07T20:10:04	COMPLETED	0:0
721813	bwa_mem_ACC5339A4	developme+	core	36	1-08:49:35	01:13:44	2020-11-07T18:54:17	2020-11-07T20:08:01	COMPLETED	0:0
721814	bwa_mem_ACC5339A4	developme+	core	36	1-09:40:37	01:17:59	2020-11-07T18:54:17	2020-11-07T20:12:16	COMPLETED	0:0
721815	bwa_mem_ACC5339A4	developme+	core	36	1-09:11:47	01:14:30	2020-11-07T18:54:17	2020-11-07T20:08:47	COMPLETED	0:0
721816	samtools_merge_ACC5339A2	developme+	core	13	01:36:07	00:08:12	2020-11-07T20:13:54	2020-11-07T20:22:06	COMPLETED	0:0
721817	samtools_merge_ACC5339A3	developme+	core	13	01:38:37	00:08:28	2020-11-07T20:18:12	2020-11-07T20:26:40	COMPLETED	0:0
721818	samtools_merge_ACC5339A4	developme+	core	13	01:38:59	00:08:30	2020-11-07T20:12:16	2020-11-07T20:20:46	COMPLETED	0:0
721819	markduplicates_ACC5339A2	developme+	core	29	15:01:38	01:02:00	2020-11-07T20:22:06	2020-11-07T21:24:06	COMPLETED	0:0
721820	markduplicates_ACC5339A3	developme+	core	29	15:35:22	01:06:21	2020-11-07T20:26:40	2020-11-07T21:33:01	COMPLETED	0:0
721821	markduplicates_ACC5339A4	developme+	core	29	15:04:54	01:01:48	2020-11-07T20:20:46	2020-11-07T21:22:34	COMPLETED	0:0
721822	gatk_baserecalibration_ACC5339A2	developme+	core	26	16:12:11	01:32:48	2020-11-07T21:24:06	2020-11-07T22:56:54	COMPLETED	0:0
721823	gatk_baserecalibration_ACC5339A3	developme+	core	26	17:21:55	01:42:05	2020-11-07T21:33:01	2020-11-07T23:15:06	COMPLETED	0:0
721824	gatk_baserecalibration_ACC5339A4	developme+	core	26	16:31:21	01:38:04	2020-11-07T21:22:34	2020-11-07T23:00:38	COMPLETED	0:0
721825	chanjo_sexcheck_ACC5339A2	developme+	core	1	01:11.547	00:01:17	2020-11-07T22:56:54	2020-11-07T22:58:11	COMPLETED	0:0
721826	chanjo_sexcheck_ACC5339A3	developme+	core	1	01:56.034	00:02:04	2020-11-07T23:15:06	2020-11-07T23:17:10	COMPLETED	0:0
721827	chanjo_sexcheck_ACC5339A4	developme+	core	1	01:21.064	00:01:26	2020-11-07T23:00:38	2020-11-07T23:02:04	COMPLETED	0:0
721828	samtools_subsample_mt_ACC5339A2	developme+	core	1	00:27.846	00:00:38	2020-11-07T22:56:54	2020-11-07T22:57:32	COMPLETED	0:0
721829	samtools_subsample_mt_ACC5339A3	developme+	core	1	00:27.264	00:00:39	2020-11-07T23:15:06	2020-11-07T23:15:45	COMPLETED	0:0
721830	samtools_subsample_mt_ACC5339A4	developme+	core	1	00:26.733	00:00:36	2020-11-07T23:00:38	2020-11-07T23:01:14	COMPLETED	0:0
721831	tiddit_coverage_ACC5339A2	developme+	core	1	25:53.330	00:26:11	2020-11-07T22:56:54	2020-11-07T23:23:05	COMPLETED	0:0
721832	tiddit_coverage_ACC5339A3	developme+	core	1	31:06.888	00:33:16	2020-11-07T23:15:06	2020-11-07T23:48:22	COMPLETED	0:0
721833	tiddit_coverage_ACC5339A4	developme+	core	1	27:13.489	00:27:29	2020-11-07T23:00:38	2020-11-07T23:28:07	COMPLETED	0:0
721834	chromograph_cov_ACC5339A2	developme+	core	1	00:19.815	00:00:26	2020-11-07T23:23:05	2020-11-07T23:23:31	COMPLETED	0:0
721835	chromograph_cov_ACC5339A3	developme+	core	1	00:20.230	00:00:28	2020-11-07T23:48:22	2020-11-07T23:48:50	COMPLETED	0:0
721836	chromograph_cov_ACC5339A4	developme+	core	1	00:19.747	00:00:24	2020-11-07T23:28:07	2020-11-07T23:28:31	COMPLETED	0:0
721837	sambamba_depth_ACC5339A2	developme+	core	2	13:42.742	00:14:19	2020-11-07T22:56:54	2020-11-07T23:11:13	COMPLETED	0:0
721838	sambamba_depth_ACC5339A3	developme+	core	2	14:21.787	00:15:36	2020-11-07T23:15:06	2020-11-07T23:30:42	COMPLETED	0:0
721839	sambamba_depth_ACC5339A4	developme+	core	2	13:16.874	00:14:18	2020-11-07T23:00:38	2020-11-07T23:14:56	COMPLETED	0:0
721840	smncopynumbercaller_livingox	developme+	core	4	01:36.043	00:02:30	2020-11-07T23:15:06	2020-11-07T23:17:36	COMPLETED	0:0
721841	star_caller_livingox	developme+	core	4	02:15.980	00:02:56	2020-11-07T23:15:06	2020-11-07T23:18:02	COMPLETED	0:0
721842	telomerecat_ar_livingox	developme+	core	12	08:58:58	01:49:27	2020-11-07T23:15:06	2020-11-08T01:04:33	COMPLETED	0:0
721843	picardtools_collectmultiplemetrics_ACC5339A2	developme+	core	2	02:32:19	02:39:38	2020-11-07T22:56:54	2020-11-08T01:36:32	COMPLETED	0:0
721844	picardtools_collectmultiplemetrics_ACC5339A3	developme+	core	2	02:25:50	02:25:40	2020-11-07T23:15:06	2020-11-08T01:40:46	COMPLETED	0:0
721845	picardtools_collectmultiplemetrics_ACC5339A4	developme+	core	2	02:26:32	02:34:07	2020-11-07T23:00:38	2020-11-08T01:34:45	COMPLETED	0:0
721846	picardtools_collecthsmetrics_ACC5339A2	developme+	core	2	01:30:54	01:30:50	2020-11-07T22:56:54	2020-11-08T00:27:44	COMPLETED	0:0
721847	picardtools_collecthsmetrics_ACC5339A3	developme+	core	2	01:29:39	01:37:19	2020-11-07T23:15:06	2020-11-08T00:52:25	COMPLETED	0:0
721848	picardtools_collecthsmetrics_ACC5339A4	developme+	core	2	01:34:52	01:34:13	2020-11-07T23:00:38	2020-11-08T00:34:51	COMPLETED	0:0
721849	cnvnator_ar_ACC5339A2	developme+	core	13	01:05:53	00:07:07	2020-11-07T22:56:54	2020-11-07T23:04:01	COMPLETED	0:0
721850	cnvnator_ar_ACC5339A3	developme+	core	13	01:08:53	00:07:19	2020-11-07T23:15:06	2020-11-07T23:22:25	COMPLETED	0:0
721851	cnvnator_ar_ACC5339A4	developme+	core	13	01:06:56	00:07:10	2020-11-07T23:00:38	2020-11-07T23:07:48	COMPLETED	0:0
721852	expansionhunter_livingox	developme+	core	3	51:01.005	00:22:11	2020-11-07T23:15:06	2020-11-07T23:37:17	COMPLETED	0:0
721853	manta_livingox	developme+	core	36	16-14:53:+	11:53:04	2020-11-07T23:15:06	2020-11-08T11:08:10	COMPLETED	0:0
721854	tiddit_livingox	developme+	core	5	04:13:34	01:28:40	2020-11-07T23:15:06	2020-11-08T00:43:46	COMPLETED	0:0
721855	sv_combinevariantcallsets_livingox	developme+	core	1	00:55.394	00:01:23	2020-11-08T11:08:10	2020-11-08T11:09:33	COMPLETED	0:0
721856	sv_annotate_livingox	developme+	core	2	16:41.046	00:12:17	2020-11-08T11:09:33	2020-11-08T11:21:50	COMPLETED	0:0
721857	sv_varianteffectpredictor_livingox	developme+	core	25	07:14.654	00:01:05	2020-11-08T11:21:50	2020-11-08T11:22:55	COMPLETED	0:0
721858	sv_vcfparser_livingox	developme+	core	13	01:44.990	00:00:29	2020-11-08T11:22:55	2020-11-08T11:23:24	COMPLETED	0:0
721859	sv_rankvariant_livingox	developme+	core	16	00:37.314	00:00:16	2020-11-08T11:23:24	2020-11-08T11:23:40	COMPLETED	0:0
721860	sv_reformat_livingox	developme+	core	5	00:13.326	00:00:18	2020-11-08T11:23:40	2020-11-08T11:23:58	COMPLETED	0:0
721861	vcf2cytosure_ar_livingox	developme+	core	4	01:21:37	00:29:13	2020-11-08T11:23:58	2020-11-08T11:53:11	COMPLETED	0:0
721862	deepvariant_ACC5339A2	developme+	core	36	4-14:29:17	04:58:38	2020-11-07T22:56:54	2020-11-08T03:55:32	COMPLETED	0:0
721863	deepvariant_ACC5339A3	developme+	core	36	4-18:12:10	05:05:08	2020-11-07T23:15:06	2020-11-08T04:20:14	COMPLETED	0:0
721864	deepvariant_ACC5339A4	developme+	core	36	4-15:17:53	04:57:59	2020-11-07T23:00:38	2020-11-08T03:58:37	COMPLETED	0:0
721865	glnexus_merge_livingox	developme+	core	30	47:26.774	00:10:44	2020-11-08T04:20:14	2020-11-08T04:30:58	COMPLETED	0:0
721866	gatk_haplotypecaller_ACC5339A2	developme+	core	16	3-17:11:42	06:13:24	2020-11-07T22:56:54	2020-11-08T05:10:18	COMPLETED	0:0
721867	gatk_haplotypecaller_ACC5339A3	developme+	core	16	3-22:46:16	06:46:23	2020-11-07T23:15:06	2020-11-08T06:01:29	COMPLETED	0:0
721868	gatk_haplotypecaller_ACC5339A4	developme+	core	16	4-04:19:03	07:11:23	2020-11-07T23:00:38	2020-11-08T06:12:01	COMPLETED	0:0
721869	gatk_genotypegvcfs_livingox	developme+	core	3	11:57.973	00:12:57	2020-11-08T06:12:01	2020-11-08T06:24:58	COMPLETED	0:0
721870	gatk_genotypegvcfs_livingox	developme+	core	3	12:19.617	00:13:08	2020-11-08T06:12:01	2020-11-08T06:25:09	COMPLETED	0:0
721871	gatk_genotypegvcfs_livingox	developme+	core	3	10:17.374	00:11:08	2020-11-08T06:12:01	2020-11-08T06:23:09	COMPLETED	0:0
721872	gatk_genotypegvcfs_livingox	developme+	core	3	10:38.824	00:11:30	2020-11-08T06:12:01	2020-11-08T06:23:31	COMPLETED	0:0
721873	gatk_genotypegvcfs_livingox	developme+	core	3	10:26.163	00:11:54	2020-11-08T06:12:01	2020-11-08T06:23:55	COMPLETED	0:0
721874	gatk_genotypegvcfs_livingox	developme+	core	3	09:16.353	00:10:50	2020-11-08T06:12:01	2020-11-08T06:22:51	COMPLETED	0:0
721875	gatk_genotypegvcfs_livingox	developme+	core	3	09:44.483	00:11:18	2020-11-08T06:12:01	2020-11-08T06:23:19	COMPLETED	0:0
721876	gatk_genotypegvcfs_livingox	developme+	core	3	08:26.866	00:09:48	2020-11-08T06:12:01	2020-11-08T06:21:49	COMPLETED	0:0
721877	gatk_genotypegvcfs_livingox	developme+	core	3	08:21.447	00:09:36	2020-11-08T06:12:01	2020-11-08T06:21:37	COMPLETED	0:0
721878	gatk_genotypegvcfs_livingox	developme+	core	3	08:57.529	00:10:33	2020-11-08T06:12:01	2020-11-08T06:22:34	COMPLETED	0:0
721879	gatk_genotypegvcfs_livingox	developme+	core	3	07:44.993	00:08:57	2020-11-08T06:12:01	2020-11-08T06:20:58	COMPLETED	0:0
721880	gatk_genotypegvcfs_livingox	developme+	core	3	08:04.998	00:09:37	2020-11-08T06:12:01	2020-11-08T06:21:38	COMPLETED	0:0
721881	gatk_genotypegvcfs_livingox	developme+	core	3	06:22.082	00:08:19	2020-11-08T06:12:01	2020-11-08T06:20:20	COMPLETED	0:0
721882	gatk_genotypegvcfs_livingox	developme+	core	3	05:40.598	00:07:45	2020-11-08T06:12:01	2020-11-08T06:19:46	COMPLETED	0:0
721883	gatk_genotypegvcfs_livingox	developme+	core	3	05:41.131	00:07:32	2020-11-08T06:12:01	2020-11-08T06:19:33	COMPLETED	0:0
721884	gatk_genotypegvcfs_livingox	developme+	core	3	05:58.192	00:08:03	2020-11-08T06:12:02	2020-11-08T06:20:05	COMPLETED	0:0
721885	gatk_genotypegvcfs_livingox	developme+	core	3	05:34.528	00:07:25	2020-11-08T06:12:02	2020-11-08T06:19:27	COMPLETED	0:0
721886	gatk_genotypegvcfs_livingox	developme+	core	3	04:56.869	00:06:33	2020-11-08T06:12:02	2020-11-08T06:18:35	COMPLETED	0:0
721887	gatk_genotypegvcfs_livingox	developme+	core	3	05:11.945	00:07:05	2020-11-08T06:12:02	2020-11-08T06:19:07	COMPLETED	0:0
721888	gatk_genotypegvcfs_livingox	developme+	core	3	04:06.689	00:05:33	2020-11-08T06:12:02	2020-11-08T06:17:35	COMPLETED	0:0
721889	gatk_genotypegvcfs_livingox	developme+	core	3	02:42.275	00:02:26	2020-11-08T06:12:02	2020-11-08T06:14:28	COMPLETED	0:0
721890	gatk_genotypegvcfs_livingox	developme+	core	3	02:51.425	00:02:33	2020-11-08T06:12:02	2020-11-08T06:14:35	COMPLETED	0:0
721891	gatk_genotypegvcfs_livingox	developme+	core	3	30:58.299	00:32:23	2020-11-08T06:12:02	2020-11-08T06:44:25	COMPLETED	0:0
721892	gatk_genotypegvcfs_livingox	developme+	core	3	04:16.910	00:04:03	2020-11-08T06:12:02	2020-11-08T06:16:05	COMPLETED	0:0
721893	gatk_genotypegvcfs_livingox	developme+	core	3	00:18.695	00:00:22	2020-11-08T06:12:02	2020-11-08T06:12:24	COMPLETED	0:0
721894	gatk_gathervcfs_livingox	developme+	core	2	01:21.046	00:01:15	2020-11-08T06:44:25	2020-11-08T06:45:40	COMPLETED	0:0
721895	gatk_variantrecalibration_livingox	developme+	core	6	43:37.794	00:40:32	2020-11-08T06:45:40	2020-11-08T07:26:12	COMPLETED	0:0
721896	gatk_combinevariantcallsets_livingox	developme+	core	5	00:06:44	00:06:04	2020-11-08T07:26:12	2020-11-08T07:32:16	COMPLETED	0:0
721897	peddy_ar_livingox	developme+	core	4	01:11.340	00:00:55	2020-11-08T07:32:16	2020-11-08T07:33:11	COMPLETED	0:0
721898	plink_livingox	developme+	core	2	01:34.013	00:01:38	2020-11-08T07:32:16	2020-11-08T07:33:54	COMPLETED	0:0
721899	rtg_vcfeval_livingox	developme+	core	36	10:05.085	00:02:52	2020-11-08T07:32:16	2020-11-08T07:35:08	COMPLETED	0:0
721900	gatk_variantevalall_ACC5339A2	developme+	core	1	04:55.111	00:05:04	2020-11-08T07:32:16	2020-11-08T07:37:20	COMPLETED	0:0
721901	gatk_variantevalall_ACC5339A3	developme+	core	1	05:16.486	00:05:29	2020-11-08T07:32:16	2020-11-08T07:37:45	COMPLETED	0:0
721902	gatk_variantevalall_ACC5339A4	developme+	core	1	05:16.569	00:05:29	2020-11-08T07:32:16	2020-11-08T07:37:45	COMPLETED	0:0
721903	prepareforvariantannotationblock_livingox	developme+	core	25	01:25.144	00:00:43	2020-11-08T07:32:16	2020-11-08T07:32:59	COMPLETED	0:0
721904	rhocall_ar_livingox	developme+	core	13	04:06.545	00:00:30	2020-11-08T07:32:59	2020-11-08T07:33:29	COMPLETED	0:0
721905	vt_ar_livingox	developme+	core	13	03:50.271	00:00:32	2020-11-08T07:33:29	2020-11-08T07:34:01	COMPLETED	0:0
721906	variant_annotation_livingox	developme+	core	13	04:29:15	00:21:53	2020-11-08T07:34:01	2020-11-08T07:55:54	COMPLETED	0:0
721907	rhocall_viz_ACC5339A2	developme+	core	1	02:59.561	00:03:23	2020-11-08T07:55:54	2020-11-08T07:59:17	COMPLETED	0:0
721908	rhocall_viz_ACC5339A3	developme+	core	1	02:58.894	00:03:23	2020-11-08T07:55:54	2020-11-08T07:59:17	COMPLETED	0:0
721909	rhocall_viz_ACC5339A4	developme+	core	1	02:57.930	00:03:23	2020-11-08T07:55:54	2020-11-08T07:59:17	COMPLETED	0:0
721910	upd_ar_ACC5339A4	developme+	core	1	07:51.370	00:08:08	2020-11-08T07:55:54	2020-11-08T08:04:02	COMPLETED	0:0
721911	chromograph_upd_ACC5339A4	developme+	core	1	02:22.572	00:02:32	2020-11-08T08:04:02	2020-11-08T08:06:34	COMPLETED	0:0
721912	frequency_filter_livingox	developme+	core	13	01:43.436	00:00:19	2020-11-08T07:55:54	2020-11-08T07:56:13	COMPLETED	0:0
721913	cadd_ar_livingox	developme+	core	16	16:46:02	01:30:32	2020-11-08T07:56:13	2020-11-08T09:26:45	COMPLETED	0:0
721914	varianteffectpredictor_livingox	developme+	core	25	19:20:49	00:50:27	2020-11-08T09:26:45	2020-11-08T10:17:12	COMPLETED	0:0
721915	vcfparser_ar_livingox	developme+	core	25	01:16:21	00:06:14	2020-11-08T10:17:12	2020-11-08T10:23:26	COMPLETED	0:0
721916	rankvariant_livingox	developme+	core	36	02:25:50	00:09:18	2020-11-08T10:23:26	2020-11-08T10:32:44	COMPLETED	0:0
721917	endvariantannotationblock_livingox	developme+	core	2	03:18.858	00:03:25	2020-11-08T10:32:45	2020-11-08T10:36:10	COMPLETED	0:0
721918	gatk_variantevalexome_ACC5339A2	developme+	core	1	02:53.865	00:03:01	2020-11-08T10:36:10	2020-11-08T10:39:11	COMPLETED	0:0
721919	gatk_variantevalexome_ACC5339A3	developme+	core	1	02:51.262	00:02:58	2020-11-08T10:36:10	2020-11-08T10:39:08	COMPLETED	0:0
721920	gatk_variantevalexome_ACC5339A4	developme+	core	1	02:49.788	00:02:58	2020-11-08T10:36:10	2020-11-08T10:39:08	COMPLETED	0:0
721921	qccollect_ar_livingox	developme+	core	1	00:14.829	00:00:18	2020-11-08T11:53:11	2020-11-08T11:53:29	COMPLETED	0:0
721922	version_collect_ar_livingox	developme+	core	1	00:32.761	00:00:51	2020-11-08T11:53:29	2020-11-08T11:54:20	COMPLETED	0:0
721923	multiqc_ar_livingox	developme+	core	1	03:39.435	00:04:50	2020-11-08T11:54:20	2020-11-08T11:59:10	COMPLETED	0:0
721924	analysisrunstatus_livingox	developme+	core	1	00:30.418	00:00:43	2020-11-08T11:59:10	2020-11-08T11:59:53	FAILED	1:0
```

Analysisrunstatus fails on:
```
Could not find /home/henrik.stranneheim/develop/livingox/analysis/721924/livingox_glnex.vcf
```
not related to this PR

**Review by ramprasadn — APPROVED** [review]
Looks good to me!

File: lib/MIP/Recipes/Analysis/Bwa_mem.pm [hunk 2]
```diff
@@ -1357,9 +1351,15 @@ sub _add_percentage_mapped_reads_from_samtools {
...
+    use MIP::Environment::Executable qw{ get_executable_base_command };
+
     ## Add percentage mapped reads to samtools stats output
+
+    my @commands = ( get_executable_base_command( { base_command => q{perl}, } ), );
+
     # Execute perl
-    print {$filehandle} q?perl -ne '?;
+    print {$filehandle} join $SPACE, @commands;
+    print {$filehandle} q? -ne '?;
```
Comment (jemten) [inline lib/MIP/Recipes/Analysis/Bwa_mem.pm:1362]: One day we should make this utility sub (and subs like it) use the perl language module instead

File: lib/MIP/Script/Setup_script.pm [hunk 3]
```diff
@@ -711,12 +748,6 @@ sub setup_script {
             store       => \$set_pipefail,
             strict_type => 1,
         },
-        sleep => {
-            allow       => [ 0, 1 ],
-            default     => 0,
-            store       => \$sleep,
-            strict_type => 1,
-        },
```
Comment (jemten) [inline lib/MIP/Script/Setup_script.pm:719]: I had forgotten about this one. It was to handle to many simultaneous calls to conda right? So nice to get rid of 👍

File: templates/mip_install_config.yaml [hunk 4]
```diff
@@ -39,6 +39,8 @@ container:
       bedtools:
     uri: docker.io/clinicalgenomics/bedtools:2.2.29
   cadd:
+    bind_path:
+      CADD.sh: reference_dir!/CADD-scripts/data/annotations:/opt/conda/share/CADD-scripts/data/annotations
```
Comment (jemten) [inline templates/mip_install_config.yaml:43]: 👍

**Review by jemten — APPROVED** [review]
This looks great! What a herculean effort to change all the program modules and recipes 💪 
Looking forward to seeing this in production. 

I take it that the local readonly constant $BASE_COMMAND is created when there are multiple programs using the same binary, like gatk? It looks like it's used a bit inconsistent at the moment. It doesn't matter though. 

Looks like the option 'sleep' should be removed from t/set_script_shell_attributes.t. Maybe the sub gnu_sleep could be removed all together now?

File: lib/MIP/Environment/Container.pm [hunk 5]
```diff
@@ -27,7 +27,88 @@ BEGIN {
...
+            required    => 1,
+            store       => \$parameter_href,
+            strict_type => 1,
+        },
+    };
+
+    check( $tmpl, $arg_href, 1 ) or croak q{Could not parse arguments!};
+
+    use MIP::Active_parameter qw{add_recipe_bind_paths};
+    use MIP::Constants qw{@CONTAINER_BIND_PATHS};
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:66]:
```suggestion
```

File: lib/MIP/Environment/Container.pm [hunk 6]
```diff
@@ -27,7 +27,88 @@ BEGIN {
...
+
+    my %recipe_executable_bind_path;
+    my @recipes = get_cache(
+        {
+            parameter_href => $parameter_href,
+            parameter_name => q{recipe},
+        }
+    );
+
+    foreach my $recipe_name (@recipes) {
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:85]:
```suggestion
    RECIPE:
    foreach my $recipe_name (@recipes) {
```

File: lib/MIP/Environment/Container.pm [hunk 7]
```diff
@@ -27,7 +27,88 @@ BEGIN {
...
+
+        my @recipe_executables = get_parameter_attribute(
+            {
+                attribute      => q{program_executables},
+                parameter_href => $parameter_href,
+                parameter_name => $recipe_name,
+            }
+        );
+        foreach my $executable (@recipe_executables) {
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:94]:
```suggestion
        RECIPE_EXECUTABLE:
        foreach my $executable (@recipe_executables) {
```

File: lib/MIP/Environment/Container.pm [hunk 8]
```diff
@@ -352,4 +497,123 @@ sub run_container {
...
+            defined     => 1,
+            required    => 1,
+            store       => \$parameter_href,
+            strict_type => 1,
+        },
+    };
+
+    check( $tmpl, $arg_href, 1 ) or croak q{Could not parse arguments!};
+
+    use MIP::Constants qw{ @CONTAINER_BIND_PATHS };
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:549]:
```suggestion
```

File: templates/mip_install_config.yaml (hunk 4 above)
Comment (henrikstranneheim) [reply templates/mip_install_config.yaml:43]: Yeah, I think that will be great to keep all the container specifics there

File: lib/MIP/Script/Setup_script.pm (hunk 3 above)
Comment (henrikstranneheim) [reply lib/MIP/Script/Setup_script.pm:719]: Yes, it was always an ugly hack. 👍

File: lib/MIP/Recipes/Analysis/Bwa_mem.pm (hunk 2 above)
Comment (henrikstranneheim) [reply lib/MIP/Recipes/Analysis/Bwa_mem.pm:1362]: Agreed! A separate refactor PR for each one should be done.

File: lib/MIP/Program/Chanjo.pm [hunk 9]
```diff
@@ -25,12 +25,14 @@ BEGIN {
...
 
     # Set the version for version checking
-    our $VERSION = 1.04;
+    our $VERSION = 1.05;
 
     # Functions and variables which can be optionally exported
     our @EXPORT_OK = qw{ chanjo_sex };
 }
 
+Readonly my $BASE_COMMAND => q{chanjo};
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply lib/MIP/Program/Chanjo.pm:34]: It was the first one I did 😄 . The general idea is that if the program actually has a base command (like chanjo) we should show and use this by using a base_command constant. However, I might have missed some of the programs. Let's fix the inconsistencies as we go along.

File: lib/MIP/Environment/Container.pm (hunk 8 above)
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply lib/MIP/Environment/Container.pm]: Fixed

**Pull request merged**